pkgs/fc/agent/fc/util/slurm.py

```diff
@@ -0,0 +1,743 @@
+        for node_name in nodes_to_wait_for:
+            node_info = get_node_info(node_name)
+
+            if is_node_drained(log.bind(time=ii), node_info):
+                log.info(
+                    "node-drained",
+                    _replace_msg="{node} is now fully drained.",
+                    node=node_name,
+                )
+
+                drained_nodes.add(node_name)
+
+        for node_name in drained_nodes:
+            nodes_to_wait_for.remove(node_name)
+
+        if not nodes_to_wait_for:
+            log.info(
+                "drain-many-finished",
+                _replace_msg="All nodes are drained after {time} iterations.",
+                time=ii,
+            )
+            return
+
+        log.debug(
+            "drain-all-wait",
+            waiting_for=ii,
+            timeout=timeout,
+            num_waiting_nodes=len(nodes_to_wait_for),
+        )
+        time.sleep(1)
+
+    # Loop finished => time limit reached
+
+    remaining_node_states = {
+        o: get_node_info(o)["state"] for o in nodes_to_wait_for
+    }
+
+    log.error(
+        "drain-many-timeout",
+        timeout=timeout,
+        node_states=remaining_node_states,
+        num_remaining=len(nodes_to_wait_for),
+        _replace_msg=(
+            "{num_waiting_nodes} node(s) did not drain in time, waited "
+            "{timeout} iterations."
+        ),
+    )
+    raise NodeStateTimeout(remaining_node_states)
+
+
+def down(log, node_name, reason: str, strict_state_check: bool = False):
+    log = log.bind(node=node_name)
+    log.debug("down-start")
+
+    node_info = get_node_info(node_name)
+    state, *flags = node_info["state"].split("+")
+    log.debug("down-state-pre", state=state, flags=flags)
+
+    if state in ("DOWN", "DOWN*"):
+        if strict_state_check:
+            log.error("down-state-error", state=state, flags=flags)
+            raise NodeStateError(state, flags)
+        else:
+            log.info(
+                "down-already-reached",
+                _replace_msg="{node} is already in {state} state.",
+                state=state,
+            )
+            return
+
+    state_change_down = {
+        "node_names": node_name,
+        "node_state": pyslurm.NODE_STATE_DOWN,
+        "reason": reason,
+    }
+    result = update_nodes(state_change_down)
+    log.debug("node-update-result", result=result)
+    log.info(
+        "down-finished",
+        _replace_msg="{node} is set to DOWN.",
+    )
+
+
+class ReadyPreCheckResult(NamedTuple):
+    state: str
+    flags: list[str]
+    action: bool
+
+
+def run_ready_pre_checks(
+    log,
+    node_name,
+    strict_state_check,
+    reason_must_match,
+    skip_in_maintenance,
+    directory,
+):
+    log = log.bind(node=node_name)
+    node_info = get_node_info(node_name)
+    state, *flags = node_info["state"].split("+")
+    log.debug("ready-pre-node-state", state=state, flags=flags)
+
+    if state in ("ALLOCATED", "IDLE", "MIXED"):
```

Comment:
This whole section is hard to swallow/consume as a human the same way as above. I think we can also use the object model to make this more expressive. Currently my brain really doesn't want to parse this long page of nested ifs sprinkled with logging code.